### PR TITLE
ARROW-4216: [Python] Add CUDA API docs

### PR DESCRIPTION
**Pull request by pitrou**
Also reorganize the API docs into several documents, and add/improve docstrings.

To allow building the docs without CUDA enabled, I added some conditional inclusion logic.  When CUDA isn't enabled, the API docs are still generated but the docstrings are empty.  This seems to be the only sane setting that doesn't produce Sphinx errors, one way or the other.

### CODE REVIEW

**Comment by pitrou** [comment]
The rendered output can be viewed here: https://pitrou.net/arrowdevdoc/python/api.html

File: docs/source/conf.py [hunk 1]
```diff
@@ -387,3 +391,32 @@
...
+try:
+    import pyarrow.cuda
+    cuda_enabled = True
+except ImportError:
+    cuda_enabled = False
+    # Mock pyarrow.cuda to avoid autodoc warnings.
+    # XXX I can't get autodoc_mock_imports to work, so mock manually instead
+    # (https://github.com/sphinx-doc/sphinx/issues/2174#issuecomment-453177550)
+    from unittest import mock
+    pyarrow.cuda = sys.modules['pyarrow.cuda'] = mock.Mock()
```
Comment (wesm) [inline docs/source/conf.py]: Gosh! Seems like this should be easier

File: python/pyarrow/array.pxi [hunk 2]
```diff
@@ -842,98 +855,147 @@ cdef wrap_array_output(PyObject* output):
...
-    pass
+    """
+    A base class for Arrow floating-point arrays.
+    """
 
 
 cdef class Int8Array(IntegerArray):
-    pass
+    """
+    Concrete class for Arrow arrays of int8 data type.
```
Comment (wesm) [inline python/pyarrow/array.pxi]: Too bad Cython doesn't have some kind of macro system.

File: python/pyarrow/io.pxi [hunk 3]
```diff
@@ -44,6 +44,15 @@ cdef extern from "Python.h":
 
 
 cdef class NativeFile:
+    """
+    The base class for all Arrow streams.
+
+    Streams are either readable or writable.  They optionally support seeking.
```
Comment (wesm) [inline python/pyarrow/io.pxi]: Though readability/writability aren't mutually exclusive (I guess the only implementation that supports both is MemoryMappedFile)

**Review by wesm — APPROVED** [review]
+1. Thank you for doing all these documentation improvements. This makes the API docs a lot nicer

File: python/pyarrow/io.pxi (hunk 3 above)
Comment (pitrou) [reply python/pyarrow/io.pxi]: Will fix :-)

**Comment by xhochy** [comment]
@kszucs @pitrou Would it be possible to build with CUDA enabled in the docker-compose setup? This probably would just depend on setting up the right SDK, one doesn't need any CUDA enabled HW?

**Comment by pitrou** [comment]
> This probably would just depend on setting up the right SDK, one doesn't need any CUDA enabled HW?

Probably indeed.

**Comment by wesm** [comment]
I don't know what happens when you try to load a shared library that depends on the CUDA runtime libraries on a machine that doesn't have a device. There is nvidia-docker which enables GPUs to be used in Docker containers. We can try and see what issues we run into...

**Comment by kou** [comment]
We can build Arrow C++ with `-DARROW_CUDA=ON` without GPU device.
.deb packages for CUDA support are built by Docker without GPU device. We just need to install `nvidia-cuda-toolkit` package to build CUDA support: https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apt/debian-stretch/Dockerfile#L51

**Comment by pitrou** [comment]
Nice. The CUDA toolkit is a huge download, though, so better make it optional.
(also, building with CUDA without a GPU device will probably fail some tests)

**Pull request closed without merging**